[Merged by Bors] - feat(LinearAlgebra/AffineSpace): affine maps/equivs determined by values on spanning sets

jessealama opened the pull request:
This PR adds extensionality lemmas showing that affine maps and equivalences are uniquely determined by their values on any set that affinely spans the entire space.

All theorems are in `Mathlib.LinearAlgebra.AffineSpace.AffineSubspace.Basic`:

- `AffineMap.linear_eqOn_vectorSpan`: If two affine maps agree on a set, their linear parts agree on the vector span
- `AffineMap.eqOn_affineSpan`: Two affine maps which agree on a set, agree on its affine span
- `AffineMap.ext_on`: If two affine maps agree on a set that spans the entire space, they are equal
- `AffineEquiv.ext_on`: If two affine equivalences agree on a set that spans the entire space, they are equal (generalized to work between different spaces)

Extracted from #30854.

jessealama commented on Mathlib/LinearAlgebra/AffineSpace/Combination.lean:
```diff
@@ -950,4 +950,43 @@ def weightedVSubOfPoint (w : ι → k) : (ι → P) × P →ᵃ[k] V where
+    {ι : Type*} (p : ι → P₁)
+    (h_span : affineSpan k (Set.range p) = ⊤)
+    (f g : P₁ →ᵃ[k] P₂)
+    (h_agree : ∀ i, f (p i) = g (p i)) : f = g := by
```
Ah, quite right -- I thought I'd taken care of that. I've switched things to use sets rather than an index type, in both lemmas.

jessealama committed:
feat(LinearAlgebra/AffineSpace): affine maps determined by values on spanning sets

Affine maps and affine equivalences are determined by their values on any
spanning set.

jessealama committed:
refactor: use Set.restrict for concision

ocfnash commented on Mathlib/LinearAlgebra/AffineSpace/Combination.lean:
```diff
@@ -950,4 +950,46 @@ def weightedVSubOfPoint (w : ι → k) : (ι → P) × P →ᵃ[k] V where
+  apply Finset.affineCombination_congr
+  · simp
+  · intro i _
+    exact congr_fun h_agree i
```
I have some minor comments as follows:

1. I think we should reuse `variable`s to save introducing `V₁`, `P₁` etc.
2. We have the equivalent result for linear maps and it is called [LinearMap.ext_on](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Span/Basic.html#LinearMap.ext_on) so I think we should follow that naming.
3. I think we should also follow the spelling of the agreement hypothesis in `LinearMap.ext_on` and thus use `Set.EqOn` here.
4. If we factor this map through `LinearMap.ext_on`, we can avoid depending on `affineCombination` and thus can move this lemma to an earlier file.

Bearing in mind all these suggestions I think we could tweak this lemma to something like the following and place it somewhere in `LinearAlgebra.AffineSpace.AffineSubspace.Basic`:
```suggestion
theorem ext_on {V₂ P₂ : Type*} [AddCommGroup V₂] [Module k V₂] [AddTorsor V₂ P₂]
    {s : Set P} {f g : P →ᵃ[k] P₂}
    (h_span : affineSpan k s = ⊤)
    (h_agree : s.EqOn f g) : f = g := by
  have aux : Set.EqOn f.linear g.linear (affineSpan k s).direction := by
    simp only [direction_affineSpan, vectorSpan_def]
    apply LinearMap.eqOn_span
    rintro - ⟨x, hx, y, hy, rfl⟩
    simp [h_agree hx, h_agree hy]
  obtain ⟨q, hq⟩ : s.Nonempty := by contrapose! h_span; simp [h_span]
  refine AffineMap.ext_linear (LinearMap.ext_on ?_ aux) (h_agree hq)
  simpa [direction_affineSpan] using
    AffineSubspace.vectorSpan_eq_top_of_affineSpan_eq_top k V P h_span
```

ocfnash commented:
Thanks, it's great to get gaps like this in our API filled!

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/Combination.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

jessealama committed:
refactor: move AffineMap.ext_on to AffineSubspace.Basic

Thanks @ocfnash for the suggestion!

ocfnash commented on Mathlib/LinearAlgebra/AffineSpace/Combination.lean:
```diff
@@ -951,3 +951,21 @@ def weightedVSubOfPoint (w : ι → k) : (ι → P) × P →ᵃ[k] V where
+    (h_agree : s.restrict T₁ = s.restrict T₂) : T₁ = T₂ := by
+  rw [← AffineEquiv.toAffineMap_inj]
+  apply AffineMap.ext_on h_span
+  exact fun x hx => congr_fun h_agree ⟨x, hx⟩
```
Similar to the other lemma, I think we should:

- move this to `Mathlib.LinearAlgebra.AffineSpace.AffineSubspace.Basic`
- rename it to `AffineEquiv.ext_on`
- state it using `Set.eqOn`
- generalise it to an equivalence between two different spaces

It will then end up looking like:
```lean
theorem ext_on {s : Set P₁} (h_span : affineSpan k s = ⊤)
    (T₁ T₂ : P₁ ≃ᵃ[k] P₂) (h_agree : s.EqOn T₁ T₂) : T₁ = T₂ :=
  (AffineEquiv.toAffineMap_inj).mp <| AffineMap.ext_on h_span h_agree
```

ocfnash reviewed (commented):
Thanks, I have one final round of minor suggestions!

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/AffineSubspace/Basic.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

jessealama committed:
refactor: generalize and move AffineEquiv.ext_on to Basic.lean

Move the affine equivalence extensionality theorem from Combination.lean to
Basic.lean, where it belongs alongside AffineMap.ext_on. The theorem is now
more general:
- Renamed from `ext_of_span_eq_top` to `ext_on` (matching AffineMap.ext_on)
- Generalized to work with different spaces (P₁ ≃ᵃ[k] P₂) instead of just
  automorphisms (P ≃ᵃ[k] P)
- Uses Set.EqOn directly instead of Set.restrict for cleaner statement
- Simplified proof using composition operator

Addresses review comment at:
https://github.com/leanprover-community/mathlib4/pull/30949#pullrequestreview-3415857572

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes
